### PR TITLE
Refactor SdkMeterProvider with Inner Structure for Better Lifecycle Control

### LINKED ISSUE #1661: [Bug]: Global metrics provider gets shutdown when local provider is dropped
### What happened?

https://github.com/open-telemetry/opentelemetry-rust/pull/1623 introduced `Drop` implementation for `SdkMeterProvider` for proper shutdown semantics. This however introduced a bug with the preferred usage pattern like it is demonstrated in the [demo](https://github.com/open-telemetry/opentelemetry-rust/blob/main/opentelemetry-otlp/examples/basic-otlp/src/main.rs).

Here's what I assume happens: when building the metrics provider [we register a copy of the provider](https://github.com/open-telemetry/opentelemetry-rust/blob/d5392dc1df13b51553577a51d736eac9b29e7125/opentelemetry-otlp/src/metric.rs#L244) as the global metrics provider. The `SdkMeterProvider` holds references to the underlying data like the pipelines in an `Arc`, so this behaviour is all good. However, when *any* provider (like the original provider) is dropped (like it happens in the example code and which seems to be the expected pattern because it is encouraged to use the global metrics provider) and since `Drop` is implemented on the `SdkMeterProvider` will call `shutdown()` which will terminate the associated pipelines which are shared with the registered global metrics provider. Hence, we see no metrics and get an error when the global metrics provider gets dropped at application end.

### API Version

0.22

### SDK Version

0.22.1

### What Exporters are you seeing the problem on?

OTLP

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by lalitb**
Fixes #1661
Design discussion issue (if applicable) #1661 

## Changes

- Refactor `SdkMeterProvider` by encapsulating its core functionality in a new `SdkMeterProviderInner` structure. 
- Drop trait now on `SdkMeterProviderInner` to ensure cleanup only occurs after the last reference is dropped.
- Added unit test to validate that shutdown through `Drop` trait is triggered only when all `SdkMeterProvider` clones are dropped.

## Merge requirement checklist

* [x] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-rust/blob/main/CONTRIBUTING.md) guidelines followed
* [x] Unit tests added/updated (if applicable)
* [ ] Appropriate `CHANGELOG.md` files updated for non-trivial, user-facing changes
* [x] Changes in public API reviewed (if applicable)

### CODE REVIEW

File: opentelemetry-sdk/src/metrics/meter_provider.rs [hunk 1]
```diff
@@ -27,6 +27,11 @@ use super::{meter::SdkMeter, pipeline::Pipelines, reader::MetricReader, view::Vi
...
 #[derive(Clone, Debug)]
 pub struct SdkMeterProvider {
+    inner: Arc<SdkMeterProviderInner>,
+}
+
+#[derive(Clone, Debug)]
+struct SdkMeterProviderInner {
     pipes: Arc<Pipelines>,
     meters: Arc<Mutex<HashMap<Scope, Arc<SdkMeter>>>>,
     is_shutdown: Arc<AtomicBool>,
```
[the remark this answers is not in the document]
Comment (lalitb) [reply opentelemetry-sdk/src/metrics/meter_provider.rs:37]: Yes, as this allows the individual meter provider fields to be shared and accessed independently of the rest of `SdkMeterProviderInner`. In our case, the [SdkMeter](https://github.com/open-telemetry/opentelemetry-rust/blob/2286378632d498ce4b2da109e5aa131b34ae1a8f/opentelemetry-sdk/src/metrics/meter_provider.rs#L30) needs to hold a reference to the `Arc<Pipelines>` without needing access to the entire meter provider.

**Comment by TommyCpp** [comment]
It may not directly related to this PR but what's our general guideline on shutdown behavior when users call `shutdown` explicitly on cloned singal providers. 

I see two possibility here:

1. If `shutdown` is called on **any** clone of the signal provider, we should shutdown the provider(but not drop resource, as the underlying resource lifetime hasn't ended)
2. `shutdown` only works when the singal provider gets dropped(essentially make it part of the `Drop` of the resource)

I want to make sure we aligned on 1

**Comment by lalitb** [comment]
> I want to make sure we aligned on 1

Yes, agree on this in case of explicit `shutdown` invoked on any of the cloned signal provider. ( and #1643 handles this for logs).

In the case when `shutdown` is not invoked, the shutdown will be done when the last cloned signal provider is dropped.  This is the existing behavior with Span and Traces, and the current PR ensures metrics too aligns with that.

**Review by cijothomas — APPROVED** [review]

**Comment by lalitb** [comment]
> What would be the usecase for a cloned signal provider except sharing which can be done via reference?

Good question, and thanks for bringing this up.

- For logs and traces, the logger, and tracer respectively maintain the clones (internally Arc) to their providers. These clones are used for accessing processors while emitting logs/spans. Reference to providers can't be used as these loggers/tracers can outlive the original scope where the provider was created.
- For Metrics, the only use-case for cloning is when the setting the global(static) meter provider, and then application also needs the handle to this provider for shutdown/flush scenario. 
 ```rust
        let provider = super::SdkMeterProvider::builder()
            .with_reader(reader.clone())
            .build();
        global::set_meter_provider(provider.clone());
         ....
         // and later
         provider.flush()
        // and eventually
         provider.shutdown()
```
 we can possibly use reference to the global meter provider (as this is the static instance), and thus restrict `Clone`, but kept it for consistency with other signals. And since this additional `Arc` redirection is not used in the hot path of measurement recording, I didn't see much concern in having it. But I am still new to Rust, and open for any ideas for simplifying this :)

**Review by TommyCpp — APPROVED** [review]

**Pull request merged**